bigbluebutton/greenlight
# GLv3-3202-backend

Backend:
             - Set the API endpoint for the signup '/api/v1/users.json'.
             - Made some changes to Rubocop.
             - Added a request spec for the users API and prepared the tests for the signup with/without valid user data.
             - Added a password confirmation examples for the User model.
             - Improved routing compatibility between SSR (Server side routing) and CSR (Client side routing).
             - Tried defining an approach for error and request handling.
             - Removed /ruby-build.
	     - Improved API resiliency by introducing a much stricter API:
            	* API calls has to have an explicit `.json` format along with an `Accept: 'application/json'` header.
         -> The TODOs were left as comments in the code.





## Description

Allow users to signup in GLv3.

## Testing Steps

1. Pull the code.
2. Install the dependencies `bundle install && npm|yarn install`.
3. Clean the previous assets build by running `rm app/assets/builds/*` (This won't remove .keep since it's hidden).
4. Clean the database and tmp files for a better isolation by running `rails tmp:clear && rails db:schema:cache:clear && rails db:migrate:reset`
5. Run the linter and specs `bundle exec rubocop --parallel && bundle exec rspec`
6. Run `./bin/dev` to run the assets builders processes and the Puma server all at once.

To test the API:
    1. Make a POST request to `/api/v1/users.json`.
     1.1 The request has to have an `Accept: application/json` header.
     1.2. The request body should have a JSON representation with this format: 
     `{ user: { :name, :email, :password, :password_confirmation } }` .
     -> When not specifying the `.json` format the API should render back the `components#index` so that CSR handles the request.
     -> When not specifying the `Accept` header correctly the same scenario should occur.
     -> For unknown routes in the server side the CSR should take charge of handling it.
     -> Not respecting the user data required format will return a 400 BAD request with a JSON object in the response body having an errors array with a message **Invalid or Missing parameters.**
     -> Any unpermitted values given in the user data will be filtered, only `name, email ,password,...` will pass.
